crit-go/magic-gen/magic_gen.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"github.com/thoas/go-funk"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+func main() {
+	// checks aregyument correctness
+	if len(os.Args) != 3 {
+		fmt.Println("Usage: magic-gen.go path/to/magic.h path/to/magic.json")
+		os.Exit(1)
+	}
+	magic_c_header := os.Args[1]
+	magic_json := os.Args[2]
+	// opens magic file
+	infile, err := os.Open(magic_c_header)
+	check(err)
+	defer infile.Close()
+	outfile, err := os.OpenFile(magic_json, os.O_CREATE|os.O_WRONLY, 0644)
+	check(err)
+	defer outfile.Close()
+	enc := json.NewEncoder(outfile)
+	mapper := make(map[string]interface{})
+	all_magic := make(map[string]string)
+	magic := make(map[string]string)
+	by_name := make(map[string]interface{})
+	by_val := make(map[uint64]interface{})
+	scanner := bufio.NewScanner(infile)
+	scanner.Split(bufio.ScanLines)
+	// scans the file line by line
+	for scanner.Scan() {
+		split := strings.Fields(scanner.Text())
+		if len(split) < 3 {
+			continue
+		}
+
+		if !itemExists(split, "#define") {
+			continue
+		}
+
+		key := split[1]
+		value := split[2]
+		if funk.Contains(all_magic, value) {
+			value := all_magic[value]
+			all_magic[key] = value
+		} else {
+			magic[key] = value
+			all_magic[key] = value
+		}
+	}
+
+	check(scanner.Err())
+
+	for k, v := range magic {
+		if v == "0x0" || v == "1" || k == "0x0" {
+			continue
+		}
+		if strings.Contains(k, "_MAGIC") == true {
+			k = strings.Replace(k, "_MAGIC", "", -1)
+		}
+		vhexedv := hex2int(v)
+		check(err)
+		by_name[k] = vhexedv
+		by_val[vhexedv] = k
+	}
+	// by_val and by_name ampped to a master map to be converted to json
+	mapper["by_name"] = by_name
+	mapper["by_val"] = by_val
+
+	err = enc.Encode(&mapper)
+	check(err)
+}
+
+func check(e error) {
+	if e != nil {
+		panic(e)
+	}
+}
+
+// function to check if item exists in map
+func itemExists(slice interface{}, item interface{}) bool {
+
+	s := reflect.ValueOf(slice)
+	if s.Kind() != reflect.Slice {
+		panic("Invalid data-type")
+	}
+
+	return s.Index(0).Interface() == item
+}
+
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x prefix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```

Comment:
is this really necessary? Can't Go convert from 0x123e to integer?

Reply:
strconv.ParseInt considers it to be a long string thus ends with "0" or invalid syntax [playground](https://play.golang.org/p/mwtttfzROIW) thus the step of cleaning it first seems to be the only option to get around the problem.

Reply:
What about `n, err := strconv.ParseUint(s, 0, 64)`? That works for me. If you set the base to `0` it works correctly.